leader-election-impl/src/main/java/com/palantir/leader/PaxosLeadershipEventRecorder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.paxos.PaxosRoundFailureException;
+import com.palantir.paxos.PaxosValue;
+
+public class PaxosLeadershipEventRecorder implements PaxosKnowledgeEventRecorder, PaxosLeaderElectionEventRecorder {
+
+    private final String leaderId;
+    private final LeadershipEvents events;
+
+    @GuardedBy("this") private PaxosValue currentRound = null;
+    @GuardedBy("this") private boolean isLeading = false;
+
+    public static PaxosLeadershipEventRecorder create(MetricRegistry metrics, String leaderUuid) {
+        return new PaxosLeadershipEventRecorder(new LeadershipEvents(metrics), leaderUuid);
+    }
+
+    @VisibleForTesting
+    PaxosLeadershipEventRecorder(LeadershipEvents events, String leaderUuid) {
+        this.events = events;
+        this.leaderId = leaderUuid;
+    }
+
+    @Override
+    public void recordProposalAttempt(long round) {
+        events.proposedLeadershipFor(round);
+    }
+
+    @Override
+    public void recordProposalFailure(PaxosRoundFailureException e) {
+        events.proposalFailure(e);
+    }
+
+    @Override
+    public synchronized void recordRound(PaxosValue round) {
+        if (isNewRound(round)) {
+            recordNewRound(round);
+        }
+    }
+
+    private void recordNewRound(PaxosValue round) {
+        if (isLeading) {
+            events.lostLeadershipFor(currentRound);
+        }
+
+        if (isLeaderFor(round)) {
+            events.gainedLeadershipFor(round);
+        }
+
+        currentRound = round;
+        isLeading = isLeaderFor(round);
+    }
+
+    @Override
+    public synchronized void recordNotLeading(PaxosValue value) {
+        if (isSameRound(value)) {
+            if (isLeading) {
+                events.lostLeadershipFor(value);
+                isLeading = false;
+            }
+        }
+    }
+
+    @Override
+    public synchronized void recordNoQuorum(PaxosValue value) {
+        if (isSameRound(value)) {
+            events.noQuorum(value);
+        }
+    }
+
+    private boolean isNewRound(PaxosValue value) {
+        return value != null && (currentRound == null || value.getRound() > currentRound.getRound());
```

Comment:
My main concern here is that the learner doesn't necessarily learn values in increasing order of round number (e.g. if there are two quick elections and spotty network connectivity or poorly placed GCs, which has happened on internal party stack). While we do synchronize on recordRound, we can't guarantee that calls to it will be in increasing order. In other words (from our LeadershipEventRecorderTest) this is technically possible:

```java
    @Test
    public void recordingRoundsOutOfOrder() {
        recorder.recordRound(ROUND_2_NOT_LEADING);
        recorder.recordRound(ROUND_1_LEADING);

        // maybe gained for ROUND_1_LEADING and lost for ROUND_1_LEADING? currently is nothing.
    }
```

This is probably still a solid approximation, though. An alternative could be to maintain some kind of sliding window behind the current round (assuming that we don't get more than window-size rounds out of sync), though that might introduce some unwanted overhead.

Would be fine with keeping this the way it is, as long as we're aware of this possibility!